File: snippets/swift/BreezSDKExamples/Sources/Metadata.swift
//
//  Metadata.swift
//  
//
//

import Foundation
import BreezSDK

func SetPaymentMetadata(sdk: BlockingBreezServices) throws {
    // ANCHOR: set-payment-metadata
    try sdk.setPaymentMetadata(hash: "target-payment-hash", metadata: #"{"myCustomValue":true}"#)    
    // ANCHOR_END: set-payment-metadata
}

func FilterPaymentMetadata(sdk: BlockingBreezServices) -> [Payment]? {
    // ANCHOR: filter-payment-metadata
    let metadataFilters = [
        MetadataFilter(
            jsonPath: "myCustomValue",
            jsonValue: "true"
        )
    ]

    let payments = try? sdk.listPayments(
        req: ListPaymentsRequest(
            metadataFilters: metadataFilters
        )
    )
    // ANCHOR_END: filter-payment-metadata

    return payments
}

func FilterPaymentMetadataString(sdk: BlockingBreezServices) -> [Payment]? {
    // ANCHOR: filter-payment-metadata-string
    let metadataFilters = [
        MetadataFilter(
            jsonPath: "myCustomValue",
            jsonValue: #""true""#
        )
    ]
    // ANCHOR_END: filter-payment-metadata-string

    return try? sdk.listPayments(
        req: ListPaymentsRequest(
            metadataFilters: metadataFilters
        )
    )
}

func FilterPaymentMetadataObject(sdk: BlockingBreezServices) -> [Payment]? {
    // ANCHOR: filter-payment-metadata-object
    // This will *NOT* work
    var metadataFilters = [
        MetadataFilter(
            jsonPath: "myCustomValue",
            jsonValue: #"[1, 2, 3]"#
        )
    ]

    // Any of these will work
    metadataFilters = [
        MetadataFilter(
            jsonPath: "myCustomValue",
            jsonValue: #"[1,2,3]"#
        )
    ]
    // ANCHOR_END: filter-payment-metadata-object

    return try? sdk.listPayments(
        req: ListPaymentsRequest(
            metadataFilters: metadataFilters
        )
    )
}

File: snippets/kotlin_mpp_lib/shared/src/commonMain/kotlin/com/example/kotlinmpplib/Metadata.kt
package com.example.kotlinmpplib

import breez_sdk.*
class Metadata {
    fun SetPaymentMetadata(sdk: BlockingBreezServices) {
        // ANCHOR: set-payment-metadata
        try {
            sdk.setPaymentMetadata("target-payment-hash", """{"myCustomValue":true}""")
        } catch (e: Exception) {
            // Handle error
        }
        // ANCHOR_END: set-payment-metadata
    }

    fun FilterPaymentMetadata(sdk: BlockingBreezServices) {
        // ANCHOR: filter-payment-metadata
        val metadataFilters = listOf(MetadataFilter(
          jsonPath = "myCustomValue",
          jsonValue = "true"
        ))

        try {
            sdk.listPayments(
                ListPaymentsRequest(
                    metadataFilters = metadataFilters
                ))
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: filter-payment-metadata
    }

    fun FilterPaymentMetadataString(sdk: BlockingBreezServices) {
        // ANCHOR: filter-payment-metadata-string
        val metadataFilters = listOf(MetadataFilter(
          jsonPath = "customerName",
          jsonValue = "\"Satoshi Nakamoto\""
        ))
        // ANCHOR_END: filter-payment-metadata-string
    }

    fun FilterPaymentMetadataObject(sdk: BlockingBreezServices) {
        // ANCHOR: filter-payment-metadata-object
        // This will *NOT* work
        val _metadataFilters = listOf(MetadataFilter(
          jsonPath = "parent.nestedArray",
          jsonValue = "[1, 2, 3]"
        ))

        // Any of these will work
        val metadataFilters = listOf(MetadataFilter(
          jsonPath = "parent.nestedArray",
          jsonValue = "[1,2,3]"
        ))
        // ANCHOR_END: filter-payment-metadata-object
    }
}
